pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -936,16 +945,21 @@ func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {
 			for _, ea := range ss.Addresses {
 				pod := c.pods.getPodByIP(ea.IP)
 				if pod == nil {
-					// This can not happen in usual case
+					// This means, the endpoint event has arrived before pod event. This might happen because
+					// PodCache is eventually consistent. We should try to get the pod from underlying store.
+					endpointBeforePodEvents.Increment()
 					if ea.TargetRef != nil && ea.TargetRef.Kind == "Pod" {
-						log.Warnf("Endpoint without pod %s %s.%s", ea.IP, ep.Name, ep.Namespace)
-						if c.Env != nil {
-							c.Env.PushContext.Add(model.EndpointNoPod, string(hostname), nil, ea.IP)
+						pod = c.pods.getPod(ea.IP, ea.TargetRef.Name, ea.TargetRef.Namespace)
+						if pod == nil {
+							// If pod is still not availalable, this an unuusual case.
+							endpointsWithNoPods.Increment()
```

Comment:
Isn't this duplicate with Line 958

Reply:
This is metric so that we can alert on this if such a thing happens

Reply:
IC